GitUI/UserControls/RevisionGrid/Graph/Rendering/GraphRenderer.cs

```diff
@@ -0,0 +1,229 @@
+﻿using System.Drawing.Drawing2D;
+using GitCommands;
+using GitExtUtils.GitUI;
+using GitUIPluginInterfaces;
+
+namespace GitUI.UserControls.RevisionGrid.Graph.Rendering
+{
+    internal static class GraphRenderer
+    {
+        internal const int MaxLanes = RevisionGraph.MaxLanes;
+
+        public static readonly int LaneLineWidth = DpiUtil.Scale(2);
+        public static readonly int LaneWidth = DpiUtil.Scale(16);
+        public static readonly int NodeDimension = DpiUtil.Scale(10);
+
+        private const int _noLane = -10;
+
+        public static void DrawItem(Graphics g, int index, int width, int rowHeight,
+            Func<int, IRevisionGraphRow?> getSegmentsForRow,
+            RevisionGraphDrawStyleEnum revisionGraphDrawStyle,
+            ObjectId headId)
+        {
+            SmoothingMode oldSmoothingMode = g.SmoothingMode;
+            Region oldClip = g.Clip;
+
+            int top = g.RenderingOrigin.Y;
+            Rectangle laneRect = new(0, top, width, rowHeight);
+            Region newClip = new(laneRect);
+            newClip.Intersect(oldClip);
+            g.Clip = newClip;
+            g.Clear(Color.Transparent);
+
+            DrawItem();
+
+            // Restore graphics options
+            g.Clip = oldClip;
+            g.SmoothingMode = oldSmoothingMode;
+
+            return;
+
+            void DrawItem()
+            {
+                IRevisionGraphRow? currentRow = getSegmentsForRow(index);
+                if (currentRow is null)
+                {
+                    return;
+                }
+
+                IRevisionGraphRow? previousRow = getSegmentsForRow(index - 1);
+                IRevisionGraphRow? nextRow = getSegmentsForRow(index + 1);
+
+                SegmentPointsInfo p = new();
+                p.Center.Y = top + (rowHeight / 2);
+                p.Start.Y = p.Center.Y - rowHeight;
+                p.End.Y = p.Center.Y + rowHeight;
+
+                LaneInfo? currentRowRevisionLaneInfo = null;
+
+                foreach (RevisionGraphSegment revisionGraphSegment in currentRow.Segments.Reverse().OrderBy(s => s.Child.IsRelative))
+                {
+                    SegmentLanesInfo lanes = GetLanes(revisionGraphSegment, previousRow, currentRow, nextRow, li => currentRowRevisionLaneInfo = li);
+                    if (!lanes.DrawFromStart && !lanes.DrawToEnd)
+                    {
+                        continue;
+                    }
+
+                    int originX = g.RenderingOrigin.X;
+                    p.Start.X = originX + (int)((lanes.StartLane + 0.5) * LaneWidth);
+                    p.Center.X = originX + (int)((lanes.CenterLane + 0.5) * LaneWidth);
+                    p.End.X = originX + (int)((lanes.EndLane + 0.5) * LaneWidth);
+
+                    Brush laneBrush = GetBrushForLaneInfo(revisionGraphSegment.LaneInfo, revisionGraphSegment.Child.IsRelative, revisionGraphDrawStyle);
```

Comment:
What's the ownership of the brush? I.e., who's responsible for disposing it?

Reply:
This is just a reference to one of the brushes from the list `RevisionGraphLaneColor.PresetGraphBrushes`.